### PR TITLE
[Transform] Introduce `microkernel` dialect optimization passes

### PR DESCRIPTION
**Pull request by huanghaixin008**
Tracking [#297](https://github.com/intel/graph-compiler/issues/297)

This PR introduces following passed to optimize microkernel dialect runtime efficiency:

- `EarlyDispatchMicrokernel`: Dispatch microkernel during initialization time to reduce runtime cost, and merge identical microkernel dispatch meanwhile
- `InvariantMicrokernelCodeMotion`: Hoist invariant microkernel-related codes to improve performance
- `MergeBranchMicrokernelContext`: Merge and hoist identical microkernel context codes out of branch if possible, enabling further hoist

### CODE REVIEW

File: test/mlir/test/gc/cpu-runner/brgemm-simple-for.mlir [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    return
+  }
+
+  func.func @main() {
+    call @simple_brgemm() : ()->()
+    cpuruntime.printf "BRGEMM DONE\n"
+    return
+  }
+
+  // CHECK: BRGEMM DONE
```
Comment (kurapov-peter) [inline test/mlir/test/gc/cpu-runner/brgemm-simple-for.mlir:55]: Could you check for the result instead?

File: lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+    callee = tryCallOp.getCalleeAttr().getAttr();
+    if (callee != StringAttr::get(op->getContext(), DNNL_BRGEMM_DISPATCH_NAME))
+      return nullptr;
+    return tryCallOp;
+  } else if (auto tryLoadOp = dyn_cast_or_null<LLVM::LoadOp>(kernelProducer)) {
+    auto tryAddrOfOp = dyn_cast_or_null<LLVM::AddressOfOp>(
+        tryLoadOp.getOperand().getDefiningOp());
+    if (!tryAddrOfOp)
+      return nullptr;
+    return traceDispatchInGlobalCtor(module, tryAddrOfOp.getGlobalName());
```
Comment (kurapov-peter) [inline lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp:95]:
```suggestion
    if (auto tryAddrOfOp = dyn_cast_or_null<LLVM::AddressOfOp>(
        tryLoadOp.getOperand().getDefiningOp()))
        return traceDispatchInGlobalCtor(module, tryAddrOfOp.getGlobalName());
```

File: lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+  auto kernelProducer = callOp.getOperand(0).getDefiningOp();
+  // Direct producer is supposed to be either `brgemm.dispatch` or LLVM::load
+  // global Any other cases are extremely rare (mostly invalid MLIR), so
+  // considered as not found
+  if (auto tryCallOp = dyn_cast_or_null<func::CallOp>(kernelProducer)) {
+    callee = tryCallOp.getCalleeAttr().getAttr();
+    if (callee != StringAttr::get(op->getContext(), DNNL_BRGEMM_DISPATCH_NAME))
+      return nullptr;
+    return tryCallOp;
+  } else if (auto tryLoadOp = dyn_cast_or_null<LLVM::LoadOp>(kernelProducer)) {
```
Comment (kurapov-peter) [inline lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp:90]:
```suggestion
  } 
  if (auto tryLoadOp = dyn_cast_or_null<LLVM::LoadOp>(kernelProducer)) {
```

File: lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp (hunk 2 above)
Comment (kurapov-peter) [reply lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp:95]: etc.

File: lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+    }
+
+    auto defaultTileDispatch = analysis.getKernelDispatch(defaultTileOps.first);
+    if (!defaultTileDispatch)
+      return rewriter.notifyMatchFailure(op, "Cannot find kernel dispatch");
+
+    for (size_t idx = 0; idx < caseRegions.size(); idx++) {
+      auto caseTileDispatch =
+          analysis.getKernelDispatch(caseTilesOps[idx].first);
+      if (!defaultTileDispatch)
```
Comment (zhczhong) [inline lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp:263]: should be `caseTileDispatch`?

File: lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+  if (lhsDispatch.getCalleeAttr().getAttr() != dispatchNameAttr ||
+      rhsDispatch.getCalleeAttr().getAttr() != dispatchNameAttr)
+    return false;
+
+  auto lhsOperands = lhsDispatch.getOperands();
+  auto rhsOperands = rhsDispatch.getOperands();
+  assert(lhsOperands.size() == rhsOperands.size() &&
+         "Inconsistent operand size");
+  for (size_t idx = 0; idx < lhsOperands.size(); idx++) {
+    if (idx == 8) {
```
Comment (zhczhong) [inline lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp:163]: Please use a meaningful constant variable name to replace the `8`

File: lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+#include "gc/Transforms/Utils/ValueUtils.h"
+#include "oneapi/dnnl/dnnl_types.h"
+
+namespace mlir::microkernel {
+#define GEN_PASS_DEF_MERGEBRANCHMICROKERNELCONTEXT
+#include "gc/Transforms/Microkernel/MicrokernelPasses.h.inc"
+
+#define DEBUG_TYPE "merge-branch-microkernel-context"
+
+// enum BrgemmCallType { INAPPLICABLE = -1, DISPATCH, TILECFG, TILERELEASE };
```
Comment (zhczhong) [inline lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp:26]: If the code is useless, it's better to remove them

File: test/mlir/test/gc/cpu-runner/brgemm-simple-for.mlir (hunk 1 above)
Comment (huanghaixin008) [reply test/mlir/test/gc/cpu-runner/brgemm-simple-for.mlir:139]: Result correctness check added.

File: lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp (hunk 2 above)
Comment (huanghaixin008) [reply lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp]: fixed.

File: lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp (hunk 3 above)
Comment (huanghaixin008) [reply lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp]: fixed.

File: test/mlir/test/gc/cpu-runner/brgemm-simple-for.mlir [hunk 7]
```diff
@@ -0,0 +1,56 @@
+// RUN: gc-opt %s --early-dispatch-microkernel --convert-microkernel-to-dnnl-func --cse --microkernel-invariant-code-motion --convert-linalg-to-loops --convert-scf-to-cf --expand-strided-metadata --lower-affine -finalize-memref-to-llvm --convert-cpuruntime-to-llvm --convert-func-to-llvm --convert-arith-to-llvm --convert-cf-to-llvm --convert-complex-to-llvm --canonicalize --cse --reconcile-unrealized-casts --symbol-dce | gc-cpu-runner -e main -entry-point-result=void
```
[the remark this answers is not in the document]
Comment (huanghaixin008) [reply test/mlir/test/gc/cpu-runner/brgemm-simple-for.mlir:1]: microkernel passes added.

File: test/mlir/test/gc/cpu-runner/brgemm-simple-for.mlir [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    linalg.fill ins(%cst : f32) outs(%alloc_2 : memref<4x8x32x32xf32>)
+    scf.for %arg0 = %c0_index to %c4_index step %c1_index {
+      scf.for %arg1 = %c0_index to %c8_index step %c1_index {
+        %alloc_3 = memref.alloc() {alignment = 64 : i64} : memref<32x32xf32>
+        linalg.fill ins(%cst : f32) outs(%alloc_3 : memref<32x32xf32>)
+        %subview = memref.subview %alloc[%arg0, 0, 0, 0] [1, 16, 32, 32] [1, 1, 1, 1] : memref<4x16x32x32xf32> to memref<16x32x32xf32, strided<[1024, 32, 1], offset: ?>>
+        %subview_4 = memref.subview %alloc_0[%arg1, 0, 0, 0] [1, 16, 32, 32] [1, 1, 1, 1] : memref<8x16x32x32xf32> to memref<16x32x32xf32, strided<[1024, 32, 1], offset: ?>>
+        %0 = microkernel.brgemm.dispatch [32, 32, 32, 32, 32, 32, 1024, 1024] flags(stride) data_type(f32, f32) 
+        microkernel.brgemm.prologue(%0) : (i64) -> ()
+        microkernel.brgemm.execute(%0, %subview, %subview_4, %alloc_3, %c16_i64, %c0_i64) : (i64, memref<16x32x32xf32, strided<[1024, 32, 1], offset: ?>>, memref<16x32x32xf32, strided<[1024, 32, 1], offset: ?>>, memref<32x32xf32>, i64, i64) -> ()
```
[the remark this answers is not in the document]
Comment (huanghaixin008) [reply test/mlir/test/gc/cpu-runner/brgemm-simple-for.mlir:39]: This is a test for actual running, it won't print any internal logs for file check;
I've added correctness check for the running result.

File: lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp (hunk 4 above)
Comment (huanghaixin008) [reply lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp]: fixed.

File: lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp (hunk 5 above)
Comment (huanghaixin008) [reply lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp]: fixed.

File: lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp (hunk 6 above)
Comment (huanghaixin008) [reply lib/gc/Transforms/Microkernel/MergeBranchMicrokernelContext.cpp]: removed.

File: lib/gc/Transforms/Microkernel/EarlyDispatchMicrokernel.cpp [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+#include "gc/Transforms/Utils/ValueUtils.h"
+#include "oneapi/dnnl/dnnl_types.h"
+
+namespace mlir::microkernel {
+#define GEN_PASS_DEF_EARLYDISPATCHMICROKERNEL
+#include "gc/Transforms/Microkernel/MicrokernelPasses.h.inc"
+
+#define DEBUG_TYPE "early-dispatch-microkernel"
+
+static FailureOr<std::string>
```
Comment (ciyongch) [inline lib/gc/Transforms/Microkernel/EarlyDispatchMicrokernel.cpp:29]: Using string as the key for global kernel cache might not be the good option when considering the post-op fusion or m_mask stuff in the future.

File: lib/gc/Transforms/Microkernel/MicrokernelInvariantCodeMotion.cpp [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+    assert(targetInfoIter != structInfo.end());
+    // move last tilerelease to end of contextRoot, and remove all
+    // others
+    if (targetInfoIter->second.hasTilereleased) {
+      rewriter.eraseOp(op);
+    } else {
+      // auto region = opStructInfo.maxInvariantScope->getRegion(0);
+      // auto block = &region.getBlocks().front();
+      // auto enditer = block->end();
+      // rewriter.moveOpBefore(op, block, enditer);
```
Comment (ciyongch) [inline lib/gc/Transforms/Microkernel/MicrokernelInvariantCodeMotion.cpp:177]: Are these useless? Or do you have any purpose to keep them?

File: lib/gc/Transforms/Microkernel/MicrokernelInvariantCodeMotion.cpp [hunk 11]
```diff
@@ -0,0 +1,434 @@
...
+  void runOnOperation() final {
+    OpStructInfoMap structInfo;
+
+    if (failed(collectBrgemmContextStructInfo(structInfo))) {
+      signalPassFailure();
+    }
+
+    RewritePatternSet patterns(&getContext());
+    patterns.add<BrgemmTilecfgRewriter>(&getContext(), structInfo);
+    patterns.add<BrgemmTilereleaseRewriter>(&getContext(), structInfo);
```
Comment (ciyongch) [inline lib/gc/Transforms/Microkernel/MicrokernelInvariantCodeMotion.cpp:421]: Do we have any mechanism to make sure the pair usage of tile cfg and tile release?

File: lib/gc/Transforms/Microkernel/EarlyDispatchMicrokernel.cpp [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+  // TODO(haixin): Add runtime backend type to global name
+  std::stringstream ss;
+  ss << "g_dispatched_microkernel_brgemm";
+
+  auto flags = op.getFlagsAttr();
+  for (auto flag : flags) {
+    auto brgemmFlag = dyn_cast_or_null<microkernel::BrgemmFlagsAttr>(flag);
+    if (!brgemmFlag)
+      return failure();
+    if (brgemmFlag.getValue() == BrgemmFlags::LIST)
```
[the remark this answers is not in the document]
Comment (huanghaixin008) [reply lib/gc/Transforms/Microkernel/EarlyDispatchMicrokernel.cpp:43]: TODO added. And brgemm type (stride/list) added to name string.

File: lib/gc/Transforms/Microkernel/MicrokernelInvariantCodeMotion.cpp (hunk 10 above)
Comment (huanghaixin008) [reply lib/gc/Transforms/Microkernel/MicrokernelInvariantCodeMotion.cpp]: Useless comments removed.

File: lib/gc/Transforms/Microkernel/EarlyDispatchMicrokernel.cpp [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+      return failure();
+    if (brgemmFlag.getValue() == BrgemmFlags::BETA_0)
+      ss << "_init";
+  }
+
+  // M, N, K, LDA, LDB, LDC, stride_a, stride_b
+  // they are in the same order with BrgemmDispatchOp inputs
+  ArrayRef<int64_t> inputs = op.getInputsAttr().asArrayRef();
```
[the remark this answers is not in the document]
Comment (huanghaixin008) [reply lib/gc/Transforms/Microkernel/EarlyDispatchMicrokernel.cpp:61]: The order of `op.getInputsAttr()` is ensured by TD-generated codes w.r.t input order written in .td files.

File: lib/gc/Transforms/Microkernel/MicrokernelInvariantCodeMotion.cpp (hunk 11 above)
Comment (huanghaixin008) [reply lib/gc/Transforms/Microkernel/MicrokernelInvariantCodeMotion.cpp:418]: Actually we don't need to ensure pair usage, we only need to call release once at the end of context. In this pass it just simply remove all release and add one at the end of parallel scope or function.

File: lib/gc/Transforms/Microkernel/EarlyDispatchMicrokernel.cpp (hunk 9 above)
Comment (huanghaixin008) [reply lib/gc/Transforms/Microkernel/EarlyDispatchMicrokernel.cpp:29]: The string name could be lengthy but I think it's pretty self-explanatory and makes pass independent of compiler's internal state. Consider such a scenario with IR going through following pipeline: 
`EarlyDispatchMicrokernel -> SomeLoweringPassProducingNewBrgemm -> ConvertLinalgToBrgemm -> EarlyDispatchMicrokernel`
If we use global var name as cache key, we can easily dedup between first and second `EarlyDispatchMicrokernel`. I think it's hard to implement this if we keep global kernel cache as some compiler's internal state, especially under test/debug scenarios using `mlir-opt` where we might run the passes one by one in different spawns of process.

For stuff like post-op fusion and mask, we can add the attr into the name as well, with predefined format, e.g.:
`
llvm.mlir.global internal @g_mask_1 = xxxx
llvm.mlir.global internal @g_dispatched_microkernel_brgemm_..._mask{g_mask_1}_fusing_relu() ...
`

File: lib/gc/Transforms/Microkernel/EarlyDispatchMicrokernel.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (ciyongch) [reply lib/gc/Transforms/Microkernel/EarlyDispatchMicrokernel.cpp:61]: Please add a note here for future possible enhancement if we encounter the order issue.

[the remark this answers is not in the document]
Comment (huanghaixin008) [reply lib/gc/Transforms/Microkernel/EarlyDispatchMicrokernel.cpp:62]: TODO added.

**Comment by ciyongch** [comment]
Please check the failed case:
scripts/correctness.sh: line 24:  5132 Segmentation fault      (core dumped) python3 -m benchgc --verbose 0 --driver linalg --case batch_reduce_matmul --md 0:16x512x64xf32 --md 1:16x64x32xf32 --md 2:512x32xf32

**Comment by huanghaixin008** [comment]
> Please check the failed case: scripts/correctness.sh: line 24: 5132 Segmentation fault (core dumped) python3 -m benchgc --verbose 0 --driver linalg --case batch_reduce_matmul --md 0:16x512x64xf32 --md 1:16x64x32xf32 --md 2:512x32xf32

correctness check failed has been fixed.

**Review by ciyongch — APPROVED** [review]

**Review by zhczhong — APPROVED** [review]

**Review by zhczhong — APPROVED** [review]

**Pull request merged**